Index: arch/mips/kernel/proc.c
===================================================================
--- arch/mips/kernel/proc.c	(revision 1)
+++ arch/mips/kernel/proc.c	(working copy)
@@ -98,7 +98,8 @@
 	/* PR22847 - Add Broadcom models */
 	[CPU_BMIPS3300]	= "BMIPS3300",
 	[CPU_BMIPS4350]	= "BMIPS4350",
-	[CPU_BMIPS4380]	= "BMIPS4380",
+//	[CPU_BMIPS4380]	= "BMIPS4380",
+	[CPU_BMIPS4380]	= "Brcm4380",
 	[CPU_BMIPS5000]	= "BMIPS5000",
 };
 
