Index: enigma2-streamproxy/src/streamproxy.c
===================================================================
--- enigma2-streamproxy.orig/src/streamproxy.c	2010-10-30 16:08:18.000000000 +0200
+++ enigma2-streamproxy/src/streamproxy.c	2010-11-27 15:02:53.151577134 +0100
@@ -8,6 +8,7 @@
 #include <sys/ioctl.h>
 #include <linux/dvb/dmx.h>
 #include <linux/dvb/version.h>
+#include <errno.h>
 
 #define MAX_PIDS 32
 #define MAX_LINE_LENGTH 512
@@ -131,8 +132,14 @@
 			static unsigned char buffer[BSIZE];
 			int r = read(demux_fd, buffer, BSIZE);
 			if (r < 0)
-				break;
-			write(1, buffer, r);
+			{
+				if (errno != EOVERFLOW && errno != EINTR)
+					break;
+			}
+			else
+			{
+				write(1, buffer, r);
+			}
 		}
 	}
 	
