diff -Naur linux-2.6.32/lib/Makefile linux-2.6.32-fixed/lib/Makefile
--- linux-2.6.32/lib/Makefile	2010-03-03 18:01:25.000000000 +0100
+++ linux-2.6.32-fixed/lib/Makefile	2010-03-03 18:04:18.000000000 +0100
@@ -68,7 +68,7 @@
 
 lib-$(CONFIG_DECOMPRESS_GZIP) += decompress_inflate.o
 lib-$(CONFIG_DECOMPRESS_BZIP2) += decompress_bunzip2.o
-lib-$(CONFIG_DECOMPRESS_LZMA) += decompress_unlzma.o
+obj-$(CONFIG_DECOMPRESS_LZMA) += decompress_unlzma.o
 
 obj-$(CONFIG_TEXTSEARCH) += textsearch.o
 obj-$(CONFIG_TEXTSEARCH_KMP) += ts_kmp.o
