diff --git a/src/gstdvbaudiosink.c b/src/gstdvbaudiosink.c
index df50a17..e31f81d 100644
--- a/src/gstdvbaudiosink.c
+++ b/src/gstdvbaudiosink.c
@@ -308,6 +308,7 @@ gst_dvbaudiosink_init (GstDVBAudioSink *klass, GstDVBAudioSinkClass * gclass)
 	klass->fd = -1;
 	klass->dump_fd = -1;
 	klass->dump_filename = NULL;
+	klass->data_written = FALSE;
 
 	gst_base_sink_set_sync (GST_BASE_SINK(klass), FALSE);
 	gst_base_sink_set_async_enabled (GST_BASE_SINK(klass), TRUE);
@@ -420,15 +421,23 @@ gst_dvbaudiosink_get_decoder_time (GstDVBAudioSink *self)
 		gint64 cur = 0;
 		static gint64 last_pos = 0;
 
-		ioctl(self->fd, AUDIO_GET_PTS, &cur);
+		// Only ask pts from the driver when we've written data to the driver
+		if (self->data_written)
+		{
+			ioctl(self->fd, AUDIO_GET_PTS, &cur);
 
-		/* workaround until driver fixed */
-		if (cur)
-			last_pos = cur;
-		else
-			cur = last_pos;
+			if (cur)
+				last_pos = cur;
+			else
+				cur = last_pos;
 
-		cur *= 11111;
+			cur *= 11111;
+		}
+		else
+		{
+			last_pos = 0;
+			cur = 0;
+		}
 
 		return cur;
 	}
@@ -972,6 +981,8 @@ gst_dvbaudiosink_render (GstBaseSink * sink, GstBuffer * buffer)
 	ASYNC_WRITE(pes_header, pes_header_size);
 	ASYNC_WRITE(data, size);
 
+	self->data_written = TRUE;
+
 	return GST_FLOW_OK;
 poll_error:
 	{
diff --git a/src/gstdvbaudiosink.h b/src/gstdvbaudiosink.h
index f88c034..d7a9cc6 100644
--- a/src/gstdvbaudiosink.h
+++ b/src/gstdvbaudiosink.h
@@ -97,6 +97,8 @@ struct _GstDVBAudioSink
 	queue_entry_t *queue;
 
 	unsigned long long timestamp;
+
+	gboolean data_written;
 };
 
 struct _GstDVBAudioSinkClass
