--- gst-plugins-bad-0.10.20.org/gst/audioparsers/plugin.c	2010-02-15 13:47:47.000000000 +0100
+++ gst-plugins-bad-0.10.20/gst/audioparsers/plugin.c	2010-09-29 16:46:06.032700502 +0200
@@ -36,7 +36,7 @@
   ret &= gst_element_register (plugin, "amrparse",
       GST_RANK_PRIMARY + 1, GST_TYPE_AMRPARSE);
   ret &= gst_element_register (plugin, "ac3parse",
-      GST_RANK_MARGINAL, GST_TYPE_AC3_PARSE);
+      GST_RANK_PRIMARY + 2, GST_TYPE_AC3_PARSE);
   ret &= gst_element_register (plugin, "flacparse",
       GST_RANK_NONE, GST_TYPE_FLAC_PARSE);
 
