diff -Naur klibc-1.5.15-org/usr/kinit/nfsmount/main.c klibc-1.5.15/usr/kinit/nfsmount/main.c
--- klibc-1.5.15-org/usr/kinit/nfsmount/main.c	2010-08-03 17:40:45.698076536 +0200
+++ klibc-1.5.15/usr/kinit/nfsmount/main.c	2010-08-03 17:39:33.354026104 +0200
@@ -70,8 +70,10 @@
 	{"noac", ~NFS_MOUNT_NOAC, NFS_MOUNT_NOAC},
 	{"lock", ~NFS_MOUNT_NONLM, 0},
 	{"nolock", ~NFS_MOUNT_NONLM, NFS_MOUNT_NONLM},
+#if 0
 	{"acl", ~NFS_MOUNT_NOACL, 0},
 	{"noacl", ~NFS_MOUNT_NOACL, NFS_MOUNT_NOACL},
+#endif
 	{"v2", ~NFS_MOUNT_VER3, 0},
 	{"v3", ~NFS_MOUNT_VER3, NFS_MOUNT_VER3},
 	{"udp", ~NFS_MOUNT_TCP, 0},
