diff --git a/linux/drivers/media/dvb/dvb-core/dmxdev.c b/linux/drivers/media/dvb/dvb-core/dmxdev.c
index 0eeb3e9..670456a 100644
--- a/linux/drivers/media/dvb/dvb-core/dmxdev.c
+++ b/linux/drivers/media/dvb/dvb-core/dmxdev.c
@@ -762,7 +762,7 @@ static int dvb_demux_open(struct inode *inode, struct file *file)
 	mutex_init(&dmxdevfilter->mutex);
 	file->private_data = dmxdevfilter;
 
-	dvb_ringbuffer_init(&dmxdevfilter->buffer, NULL, 8192);
+	dvb_ringbuffer_init(&dmxdevfilter->buffer, NULL, 8192*16);
 	dmxdevfilter->type = DMXDEV_TYPE_NONE;
 	dvb_dmxdev_filter_state_set(dmxdevfilter, DMXDEV_STATE_ALLOCATED);
 	init_timer(&dmxdevfilter->timer);
diff --git a/linux/drivers/media/dvb/dvb-core/dvb_demux.h b/linux/drivers/media/dvb/dvb-core/dvb_demux.h
index 8567965..f0a448a 100644
--- a/linux/drivers/media/dvb/dvb-core/dvb_demux.h
+++ b/linux/drivers/media/dvb/dvb-core/dvb_demux.h
@@ -61,6 +61,11 @@ struct dvb_demux_filter {
 
 	u16 hw_handle;
 	struct timer_list timer;
+////////////////////////////////////////////////////
+	void *priv;
+	void *buf;
+	int is_recpump;
+	struct dvb_demux_filter *pre;
 };
 
 #define DMX_FEED_ENTRY(pos) list_entry(pos, struct dvb_demux_feed, list_head)
diff --git a/linux/drivers/media/dvb/dvb-core/dvb_frontend.c b/linux/drivers/media/dvb/dvb-core/dvb_frontend.c
index 9440b35..d1b642d 100644
--- a/linux/drivers/media/dvb/dvb-core/dvb_frontend.c
+++ b/linux/drivers/media/dvb/dvb-core/dvb_frontend.c
@@ -31,6 +31,7 @@
 #include <linux/wait.h>
 #include <linux/slab.h>
 #include <linux/poll.h>
+#include "compat.h"
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,26)
 #include <linux/semaphore.h>
 #endif
@@ -47,7 +48,7 @@
 
 #include "dvb_frontend.h"
 #include "dvbdev.h"
-#include "compat.h"
+
 #include <linux/dvb/version.h>
 
 static int dvb_frontend_debug;
@@ -1684,6 +1685,14 @@ static int dvb_frontend_ioctl_legacy(struct inode *inode, struct file *file,
 			err = fe->ops.read_ucblocks(fe, (__u32*) parg);
 		break;
 
+#ifndef SW_LOOP_THROUGH
+#define SW_LOOP_THROUGH	_IOW('o',103,int)
+#endif
+
+	case SW_LOOP_THROUGH:
+		if (fe->ops.set_sw_loopthrough)
+			err = fe->ops.set_sw_loopthrough(fe, (int*) parg);
+		break;
 
 	case FE_DISEQC_RESET_OVERLOAD:
 		if (fe->ops.diseqc_reset_overload) {
diff --git a/linux/drivers/media/dvb/dvb-core/dvb_frontend.h b/linux/drivers/media/dvb/dvb-core/dvb_frontend.h
index bf0e6be..c992277 100644
--- a/linux/drivers/media/dvb/dvb-core/dvb_frontend.h
+++ b/linux/drivers/media/dvb/dvb-core/dvb_frontend.h
@@ -305,6 +305,8 @@ struct dvb_frontend_ops {
 
 	int (*set_property)(struct dvb_frontend* fe, struct dtv_property* tvp);
 	int (*get_property)(struct dvb_frontend* fe, struct dtv_property* tvp);
+
+	int (*set_sw_loopthrough)(struct dvb_frontend* fe, int* arg);
 };
 
 #define MAX_EVENT 8
diff --git a/linux/drivers/media/dvb/dvb-core/dvbdev.c b/linux/drivers/media/dvb/dvb-core/dvbdev.c
index 9720ec5..715e876 100644
--- a/linux/drivers/media/dvb/dvb-core/dvbdev.c
+++ b/linux/drivers/media/dvb/dvb-core/dvbdev.c
@@ -55,11 +55,14 @@ static const char * const dnames[] = {
 #define MAX_DVB_MINORS		256
 #define DVB_MAX_IDS		MAX_DVB_MINORS
 #else
-#define DVB_MAX_IDS		4
-#define nums2minor(num,type,id)	((num << 6) | (id << 4) | type)
+#define DVB_MAX_IDS		8
+#define nums2minor(num,type,id)	((num << 7) | (id << 4) | type)
 #define MAX_DVB_MINORS		(DVB_MAX_ADAPTERS*64)
 #endif
 
+
+
+
 static struct class *dvb_class;
 
 static struct dvb_device *dvb_minors[MAX_DVB_MINORS];
