--- libmms-0.6/src/mmsh.c	2010-05-31 12:13:04.000000000 +0200
+++ libmms-0.6P/src/mmsh.c	2012-04-03 16:21:38.470101550 +0200
@@ -217,7 +217,7 @@
 static off_t fallback_io_read(void *data, int socket, char *buf, off_t num)
 {
   off_t len = 0, ret;
-/*   lprintf("%d\n", fallback_io_select(data, socket, MMS_IO_READ_READY, 1000)); */
+  if (fallback_io_select(data, socket, MMS_IO_READ_READY, 1000) <= 0) return 0;
   errno = 0;
   while (len < num)
   {
