Index: arch/mips/kernel/proc.c
===================================================================
--- arch/mips/kernel/proc.c	(revision 1)
+++ arch/mips/kernel/proc.c	(working copy)
@@ -84,7 +84,8 @@
 	[CPU_20KC]	= "MIPS 20Kc",
 	[CPU_24K]	= "MIPS 24K",
 	[CPU_25KF]	= "MIPS 25Kf",
-	[CPU_34K]	= "MIPS 34K",
+//	[CPU_34K]	= "MIPS 34K",
+	[CPU_34K]	= "Brcm4380",
 	[CPU_74K]	= "MIPS 74K",
 	[CPU_VR4111]	= "NEC VR4111",
 	[CPU_VR4121]	= "NEC VR4121",
