--- libtuxtxt/tuxtxt_common.h	2012-03-09 11:49:45.813733054 +0100
+++ libtuxtxt_orig/tuxtxt_common.h	2012-03-09 22:53:39.426628824 +0100
@@ -5604,6 +5604,10 @@
 		return 0;
 	}
+	#define PAGE_SIZE (4096)
+	if (renderinfo->fix_screeninfo.smem_start & (PAGE_SIZE-1))
+		renderinfo->lfb += (renderinfo->fix_screeninfo.smem_start & (PAGE_SIZE-1));
+
 	tuxtxt_ClearFB(renderinfo,tuxtxt_color_transp);
 	tuxtxt_ClearBB(renderinfo,tuxtxt_color_transp); /* initialize backbuffer */
 	for (i = 0; i < 40 * 25; i++)
