*** linux-2.6.22.19/arch/mips/tangox/console.orig	Sun May 22 14:20:58 2011
--- linux-2.6.22.19/arch/mips/tangox/console.c	Tue Oct  4 13:55:35 2011
***************
*** 99,106 ****
  #else
  	WR_UART_REG32(CPU_UART_CLKSEL, 0x1);
  #ifdef CONFIG_TANGO2
! 	div = (TANGO3_UART_FREQUENCY / baud) >> 4;
! 	if (((((TANGO3_UART_FREQUENCY * 10) / baud) >> 4) % 10) >= 5)
  		div++;
  #elif defined(CONFIG_TANGO3)
  	div = (TANGO3_UART_FREQUENCY / baud) >> 4;
--- 99,106 ----
  #else
  	WR_UART_REG32(CPU_UART_CLKSEL, 0x1);
  #ifdef CONFIG_TANGO2
! 	div = (TANGOX_BASE_FREQUENCY / baud) >> 4;
! 	if (((((TANGOX_BASE_FREQUENCY * 10) / baud) >> 4) % 10) >= 5)
  		div++;
  #elif defined(CONFIG_TANGO3)
  	div = (TANGO3_UART_FREQUENCY / baud) >> 4;
