From ec9ddc5d1c40b789f07d86ab25b797ba07466010 Mon Sep 17 00:00:00 2001
From: Mark Nauwelaerts <mark.nauwelaerts@collabora.co.uk>
Date: Mon, 10 May 2010 18:33:03 +0200
Subject: [PATCH] qtdemux: fix partial parsing of ctts table

Fixes #616516.
---
 gst/qtdemux/qtdemux.c |    5 +++--
 1 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/gst/qtdemux/qtdemux.c b/gst/qtdemux/qtdemux.c
index 8f89ee7..7549809 100644
--- a/gst/qtdemux/qtdemux.c
+++ b/gst/qtdemux/qtdemux.c
@@ -4676,10 +4676,11 @@ ctts:
         cur->pts_offset = ctts_soffset;
         cur++;
 
-        if (G_UNLIKELY (cur > last))
+        if (G_UNLIKELY (cur > last)) {
           /* save state */
           stream->ctts_sample_index = j + 1;
-        goto done;
+          goto done;
+        }
       }
       stream->ctts_sample_index = 0;
       stream->ctts_index++;
-- 
1.6.0.4

