diff -Naur linux-2.6.18-org/fs/Kconfig linux-2.6.18/fs/Kconfig
--- linux-2.6.18-org/fs/Kconfig	2010-01-28 17:05:14.000000000 +0100
+++ linux-2.6.18/fs/Kconfig	2010-01-28 17:07:32.000000000 +0100
@@ -781,6 +781,9 @@
 
 	  It is perfectly safe to say N here.
 
+config CDFS_FS
+	tristate "CDFS file system support"
+
 endmenu
 
 menu "Pseudo filesystems"
diff -Naur linux-2.6.18-org/fs/Makefile linux-2.6.18/fs/Makefile
--- linux-2.6.18-org/fs/Makefile	2010-01-28 17:05:14.000000000 +0100
+++ linux-2.6.18/fs/Makefile	2010-01-28 17:08:02.000000000 +0100
@@ -93,6 +93,7 @@
 obj-$(CONFIG_ADFS_FS)		+= adfs/
 obj-$(CONFIG_FUSE_FS)		+= fuse/
 obj-$(CONFIG_UDF_FS)		+= udf/
+obj-$(CONFIG_CDFS_FS)		+= cdfs/
 obj-$(CONFIG_SUN_OPENPROMFS)	+= openpromfs/
 obj-$(CONFIG_JFS_FS)		+= jfs/
 obj-$(CONFIG_XFS_FS)		+= xfs/
