diff -auNr gst-plugins-base-0.10.21/gst/subparse/Makefile.am gst-plugins-base-0.10.21_samihack/gst/subparse/Makefile.am
--- gst-plugins-base-0.10.21/gst/subparse/Makefile.am	2008-09-10 10:02:38.000000000 +0000
+++ gst-plugins-base-0.10.21_samihack/gst/subparse/Makefile.am	2008-12-09 07:48:49.000000000 +0000
@@ -1,10 +1,6 @@
 plugin_LTLIBRARIES = libgstsubparse.la
 
-if USE_XML
 SAMIPARSE_SOURCES = samiparse.c samiparse.h
-else
-SAMIPARSE_SOURCES = 
-endif
 
 libgstsubparse_la_SOURCES = \
 	gstssaparse.c \
diff -auNr gst-plugins-base-0.10.21/gst/subparse/samiparse.c gst-plugins-base-0.10.21_samihack/gst/subparse/samiparse.c
--- gst-plugins-base-0.10.21/gst/subparse/samiparse.c	2008-09-10 10:02:39.000000000 +0000
+++ gst-plugins-base-0.10.21_samihack/gst/subparse/samiparse.c	2008-12-09 07:50:06.000000000 +0000
@@ -19,6 +19,7 @@
 
 #include "samiparse.h"
 
+#ifndef GST_DISABLE_XML
 #include <libxml/HTMLparser.h>
 #include <string.h>
 
@@ -464,3 +465,33 @@
   }
   return NULL;
 }
+
+#else /* GST_DISABLE_XML */
+
+gchar *
+parse_sami (ParserState * state, const gchar * line)
+{
+  /* our template caps should not include sami in this case */
+  g_assert_not_reached ();
+}
+
+void
+sami_context_init (ParserState * state)
+{
+  return;
+}
+
+void
+sami_context_deinit (ParserState * state)
+{
+  return;
+}
+
+void
+sami_context_reset (ParserState * state)
+{
+  return;
+}
+
+#endif /* GST_DISABLE_XML */
+
