diff --git a/arch/mips/brcmstb/setup.c b/arch/mips/brcmstb/setup.c
index fcd6c52..259ec5c 100644
--- a/arch/mips/brcmstb/setup.c
+++ b/arch/mips/brcmstb/setup.c
@@ -81,12 +81,6 @@ static struct plat_serial8250_port brcm_16550_ports[] = {
 #ifdef CONFIG_BRCM_UARTA_IS_16550
 BRCM_16550_PLAT_DEVICE(BCHP_UARTA_REG_START, BRCM_IRQ_UARTA)
 #endif
-#ifdef CONFIG_BRCM_UARTB_IS_16550
-BRCM_16550_PLAT_DEVICE(BCHP_UARTB_REG_START, BRCM_IRQ_UARTB)
-#endif
-#ifdef CONFIG_BRCM_UARTC_IS_16550
-BRCM_16550_PLAT_DEVICE(BCHP_UARTC_REG_START, BRCM_IRQ_UARTC)
-#endif
 	{
 		.flags = 0,
 	}
