Index: enigma2-plugins/mytube/src/plugin.py
===================================================================
--- enigma2-plugins.orig/mytube/src/plugin.py	2010-05-15 15:26:58.818103756 +0200
+++ enigma2-plugins/mytube/src/plugin.py	2010-05-15 15:29:37.921641847 +0200
@@ -38,8 +38,26 @@
 from time import time
 
 from Screens.InfoBarGenerics import InfoBarShowHide, InfoBarSeek, InfoBarNotifications, InfoBarServiceNotifications
-from enigma import eTPM
 from __init__ import bin2long, long2bin, rsa_pub1024, decrypt_block
+
+
+
+class eTPM:
+		TPMD_DT_LEVEL3_CERT = 5
+		TPMD_DT_LEVEL2_CERT = 4
+		def __init__(self):
+			self.TPM = self
+		def getCert(self, *args):
+			return "12345678"
+		def challenge(self, r):
+			return r
+def decrypt_block(a,b):
+		return 80*"-" + a
+def get_rnd():
+		return "12345678"
+def validate_cert(*args):
+		return "12345678"
+
 etpm = eTPM()
 rootkey = ['\x9f', '|', '\xe4', 'G', '\xc9', '\xb4', '\xf4', '#', '&', '\xce', '\xb3', '\xfe', '\xda', '\xc9', 'U', '`', '\xd8', '\x8c', 's', 'o', '\x90', '\x9b', '\\', 'b', '\xc0', '\x89', '\xd1', '\x8c', '\x9e', 'J', 'T', '\xc5', 'X', '\xa1', '\xb8', '\x13', '5', 'E', '\x02', '\xc9', '\xb2', '\xe6', 't', '\x89', '\xde', '\xcd', '\x9d', '\x11', '\xdd', '\xc7', '\xf4', '\xe4', '\xe4', '\xbc', '\xdb', '\x9c', '\xea', '}', '\xad', '\xda', 't', 'r', '\x9b', '\xdc', '\xbc', '\x18', '3', '\xe7', '\xaf', '|', '\xae', '\x0c', '\xe3', '\xb5', '\x84', '\x8d', '\r', '\x8d', '\x9d', '2', '\xd0', '\xce', '\xd5', 'q', '\t', '\x84', 'c', '\xa8', ')', '\x99', '\xdc', '<', '"', 'x', '\xe8', '\x87', '\x8f', '\x02', ';', 'S', 'm', '\xd5', '\xf0', '\xa3', '_', '\xb7', 'T', '\t', '\xde', '\xa7', '\xf1', '\xc9', '\xae', '\x8a', '\xd7', '\xd2', '\xcf', '\xb2', '.', '\x13', '\xfb', '\xac', 'j', '\xdf', '\xb1', '\x1d', ':', '?']
 
