--- ~/imdb/src/plugin.py~	2009-12-07 14:33:35.000000000 +0100
+++ ~/imdb/src/plugin.py	2010-01-15 18:00:22.000000000 +0100
@@ -528,9 +528,12 @@
 		Screen.__init__(self, session)
 		self["headline"] = Label(_("IMDb Plugin"))
 
-def eventinfo(session, servicelist, **kwargs):
-	ref = session.nav.getCurrentlyPlayingServiceReference()
-	session.open(IMDBEPGSelection, ref)
+def eventinfo(session, eventName="", **kwargs):
+	if eventName != "":
+		session.open(IMDB, eventName)
+	else:
+		ref = session.nav.getCurrentlyPlayingServiceReference()
+		session.open(IMDBEPGSelection, ref)
 
 def main(session, eventName="", **kwargs):
 	session.open(IMDB, eventName)
