Index: enigma2-plugins/netcaster/src/bin/StreamInterface.py
===================================================================
--- enigma2-plugins.orig/netcaster/src/bin/StreamInterface.py
+++ enigma2-plugins/netcaster/src/bin/StreamInterface.py
@@ -103,10 +103,9 @@ class Stream:
 
     def _gotPLSContent(self, lines):
 		if lines.startswith("ICY "):
-			print "PLS expected, but got ICY stream"
+			print "[NETcaster] PLS expected, but got ICY stream"
 			self.type = "mp3"
 			self.callback(self.url)
-
 		else:
 			for line in lines.split('\n'):
 			    if line.startswith("File"):
@@ -115,9 +114,11 @@ class Stream:
 			        break
 			    print "Skipping:", line
 
-    def _errorPLSContent(self, data, callback):
-    	callback(None)
-
+    def _errorPLSContent(self, data):
+        print "[NETcaster] _errorPLSContent", data
+        print "[NETcaster] _errorPLSContent let's assume it's a stream"
+        self.type = "mp3"
+        self.callback(self.url)
 
     def setFavorite(self,TrueFalse):
         self.isfavorite = TrueFalse
Index: enigma2-plugins/netcaster/src/bin/StreamPlayer.py
===================================================================
--- enigma2-plugins.orig/netcaster/src/bin/StreamPlayer.py
+++ enigma2-plugins/netcaster/src/bin/StreamPlayer.py
@@ -10,6 +10,7 @@ class StreamPlayer:
 	    self.oldService = self.session.nav.getCurrentlyPlayingServiceReference()
 	    self.session.nav.event.append(self.__event)
 	    self.metadatachangelisteners = []
+	    self.onStop = []
 
 	def __event(self, ev):
 	    print "[NETcaster.StreamPlayer] EVENT ==>", ev
@@ -19,6 +20,11 @@ class StreamPlayer:
    				#it seems, that only Title is avaible for now
    				sTagTitle = currentServiceRef.info().getInfoString(iServiceInformation.sTagTitle)
 				self._onMetadataChanged(sTagTitle)
+# TODO: Figure out the correct event for "STOP", 1 appears to be wrong.
+#	    elif ev == 1:
+#	       for c in self.onStop:
+#	           c()
+		
 
 	def _onMetadataChanged(self,title):
 		for i in self.metadatachangelisteners:
Index: enigma2-plugins/netcaster/src/bin/plugin.py
===================================================================
--- enigma2-plugins.orig/netcaster/src/bin/plugin.py
+++ enigma2-plugins/netcaster/src/bin/plugin.py
@@ -93,6 +93,7 @@ class NETcasterScreenBrowser(Screen):
         global streamplayer
         if streamplayer is not None:
              streamplayer.metadatachangelisteners.append(self.onMetadataChanged)
+             streamplayer.onStop.append(self._onStop)
 
     def disconnectFromMetadataUpdates(self):
         global streamplayer
@@ -100,6 +101,10 @@ class NETcasterScreenBrowser(Screen):
              streamplayer.metadatachangelisteners.remove(self.onMetadataChanged)
         except Exception,e:
             pass
+        try:
+             streamplayer.onStop.remove(self._onStop)
+        except Exception,e:
+            pass
 
     def onMetadataChanged(self,title):
         try:
@@ -147,14 +152,17 @@ class NETcasterScreenBrowser(Screen):
     def yellow(self):
         pass
 
+    def _onStop(self):
+        self["pixred"].setText("")
+        self.setTitle("%s (%s)"%(myname,self.currentPlugin.nameshort))
+
     def stream_stop(self):
         global streamplayer
         if streamplayer.is_playing:
             print "[",myname,"] stream_startstop -> stop"
             streamplayer.stop()
             self.disconnectFromMetadataUpdates()
-            self["pixred"].setText("")
-            self.setTitle("%s (%s)"%(myname,self.currentPlugin.nameshort))
+            self._onStop()
 
     def stream_start(self):
         global streamplayer
