Index: enigma2-plugins/networkbrowser/src/plugin.py
===================================================================
--- enigma2-plugins.orig/networkbrowser/src/plugin.py	2009-03-15 09:07:59.000000000 +0100
+++ enigma2-plugins/networkbrowser/src/plugin.py	2010-09-04 14:53:42.892347397 +0200
@@ -25,11 +25,23 @@
 def MountManagerCallFunction(iface):
 	return MountManagerMain
 
+def RemountMain(session, iface = None, **kwargs):
+	from AutoMount import iAutoMount
+	iAutoMount.getAutoMountPoints() 
+
+def RemountCallFunction(iface):
+	if iNetwork.getAdapterAttribute(iface, "up"):
+		return RemountMain
+
 def Plugins(path, **kwargs):
 	global plugin_path
 	plugin_path = path
 	return [
 		PluginDescriptor(name=_("NetworkBrowser"), description=_("Search for network shares"), where = PluginDescriptor.WHERE_NETWORKSETUP, fnc={"ifaceSupported": NetworkBrowserCallFunction, "menuEntryName": lambda x: _("NetworkBrowser"), "menuEntryDescription": lambda x: _("Search for network shares...")}),
-		PluginDescriptor(name=_("MountManager"), description=_("Manage network shares"), where = PluginDescriptor.WHERE_NETWORKSETUP, fnc={"ifaceSupported": MountManagerCallFunction, "menuEntryName": lambda x: _("MountManager"), "menuEntryDescription": lambda x: _("Manage your network shares...")})
+		PluginDescriptor(name=_("MountManager"), description=_("Manage network shares"), where = PluginDescriptor.WHERE_NETWORKSETUP, fnc={"ifaceSupported": MountManagerCallFunction, "menuEntryName": lambda x: _("MountManager"), "menuEntryDescription": lambda x: _("Manage your network shares...")}),
+		PluginDescriptor(name=_("Mount again"), description=_("Attempt to mount shares again"), where = PluginDescriptor.WHERE_NETWORKSETUP,
+			fnc={"ifaceSupported": RemountCallFunction,
+				"menuEntryName": lambda x: _("Mount again"),
+				"menuEntryDescription": lambda x: _("Attempt to recover lost mounts (in background)")})
 	]
 
