Index: git/Makefile.am
===================================================================
--- git.orig/Makefile.am	2010-12-17 19:41:52.000000000 +0100
+++ git/Makefile.am	2010-12-23 17:13:39.532997964 +0100
@@ -3,7 +3,6 @@
 	ac3lipsync \
 	aihdcontrol \
 	antiscrollbar \
-	automatictimerlistcleanup \
 	automaticvolumeadjustment \
 	autoresolution \
 	autotimer \
@@ -23,7 +22,6 @@
 	filebrowser \
 	fritzcall \
 	ftpbrowser \
-	genuinedreambox \
 	googlemaps \
 	growlee \
 	httpproxy \
@@ -63,7 +61,6 @@
 	setpasswd \
 	shoutcast \
 	simplerss \
-	startupservice \
 	startuptostandby \
 	tageditor \
 	trafficinfo \
