Index: enigma2-plugins/webinterface/src/plugin.py
===================================================================
--- enigma2-plugins.orig/webinterface/src/plugin.py	2010-07-12 18:52:37.848238337 +0200
+++ enigma2-plugins/webinterface/src/plugin.py	2010-07-12 18:54:18.018069776 +0200
@@ -40,7 +40,7 @@
 config.plugins.Webinterface.http.auth = ConfigYesNo(default=False)
 
 config.plugins.Webinterface.https = ConfigSubsection()
-config.plugins.Webinterface.https.enabled = ConfigYesNo(default=True)
+config.plugins.Webinterface.https.enabled = ConfigYesNo(default=False)
 config.plugins.Webinterface.https.port = ConfigInteger(default = 443, limits=(1, 65535) )
 config.plugins.Webinterface.https.auth = ConfigYesNo(default=True)
 
