Index: enigma2-plugins/webinterface/src/WebComponents/Sources/Movie.py
===================================================================
--- enigma2-plugins.orig/webinterface/src/WebComponents/Sources/Movie.py
+++ enigma2-plugins/webinterface/src/WebComponents/Sources/Movie.py
@@ -70,7 +70,7 @@ class Movie(Source):
 			rtime = info.getInfo(serviceref, iServiceInformation.sTimeCreate)
 
 			if rtime > 0:
-				t = FuzzyTime(rtime)
+				t = FuzzyTime(rtime, inPast=True)
 				begin_string = t[0] + ", " + t[1]
 			else:
 				begin_string = "undefined"
