Index: enigma2-plugins/webinterface/src/WebComponents/Sources/Movie.py
===================================================================
--- enigma2-plugins.orig/webinterface/src/WebComponents/Sources/Movie.py
+++ enigma2-plugins/webinterface/src/WebComponents/Sources/Movie.py
@@ -67,6 +67,9 @@ class Movie(Source):
 		tag = self.cmd['tag']
 		tag = tag and tag.lower()
 		for (serviceref, info, begin, unknown) in self.movielist.list:
+			if serviceref.flags & eServiceReference.mustDescent:
+				# Skip subdirectories (TODO: Browse?)
+				continue
 			rtime = info.getInfo(serviceref, iServiceInformation.sTimeCreate)
 
 			if rtime > 0:
