diff -Nuar a/lib/python/Plugins/SystemPlugins/Videomode/VideoHardware.py b/lib/python/Plugins/SystemPlugins/Videomode/VideoHardware.py
--- a/lib/python/Plugins/SystemPlugins/Videomode/VideoHardware.py	2012-05-16 13:51:53.000000000 +0200
+++ b/lib/python/Plugins/SystemPlugins/Videomode/VideoHardware.py	2012-05-16 00:02:10.000000000 +0200
@@ -10,6 +10,7 @@
 # available and preferred modes, as well as handling the currently
 # selected mode. No other strict checking is done.
 class VideoHardware:
+	hw_type = HardwareInfo().get_device_name()
 	rates = { } # high-level, use selectable modes.
 
 	modes = { }  # a list of (high-level) modes for a certain port.
@@ -38,6 +39,11 @@
 								"60Hz":		{ 60: "1080i" },
 								"multi":	{ 50: "1080i50", 60: "1080i" } }
 
+	if hw_type == "me" or hw_type == "minime" :
+		rates["1080p"] =		{ "50Hz":		{ 50: "1080p50" },
+								"60Hz":		{ 60: "1080p" },
+								"multi":	{ 50: "1080p50", 60: "1080p" } }
+
 	rates["PC"] = { 
 		"1024x768": { 60: "1024x768" }, # not possible on DM7025
 		"800x600" : { 60: "800x600" },  # also not possible
@@ -56,7 +62,11 @@
 
 	modes["Scart"] = ["PAL", "NTSC", "Multi"]
 	modes["YPbPr"] = ["720p", "1080i", "576p", "480p", "576i", "480i"]
	if hw_type == "me" or hw_type == "minime" :
		modes["DVI"] = ["720p", "1080p", "1080i", "576p", "480p", "576i", "480i"]
		config.av.edid_override = True
	else:
		modes["DVI"] = ["720p", "1080i", "576p", "480p", "576i", "480i"]
 	modes["DVI-PC"] = ["PC"]
 
 	widescreen_modes = set(["720p", "1080i"])
@@ -97,10 +107,14 @@
 
 		self.readAvailableModes()
 
+		if self.hw_type == "me" or self.hw_type == "minime" : del self.modes["Scart"]
+
 		if self.modes.has_key("DVI-PC") and not self.getModeList("DVI-PC"):
 			print "remove DVI-PC because of not existing modes"
 			del self.modes["DVI-PC"]
 
+		if self.hw_type == "me" or self.hw_type == "minime" : self.readPreferredModes()
+
 		self.createConfig()
 #		self.on_hotplug.append(self.createConfig)
 
@@ -151,10 +165,13 @@
 		rate = self.rates[mode][rate]
 		for mode in rate.values():
 			# DVI modes must be in "modes_preferred"
-#			if port == "DVI":
-#				if mode not in self.modes_preferred and not config.av.edid_override.value:
-#					print "no, not preferred"
-#					return False
+			if port == "DVI":
+##### Only for test #####
+				if self.hw_type == "me" or self.hw_type == "minime" :
+					if mode not in self.modes_preferred and not config.av.edid_override.value:
+						print "no, not preferred"
+						return False
+##### Only for test #####
 			if mode not in self.modes_available:
 				return False
 		return True
